spmd/test/_utils.py

```diff
@@ -75,3 +90,162 @@ def wrapper(
+            sharding_choices = sharding_choices + [
+                Shard(i)
+                for i, s, in enumerate(arg.shape)
+                if s > 1 and s % mesh_size == 0
+            ]
+        # TODO: add multi mesh choices
+        # all_choices = itertools.product(
+        #     *(self.mesh.ndim * [sharding_choices])
+        # )
+        return sharding_choices
+
+    def __iter__(self) -> "DTensorConverter":
+        return self
+
+    def __next__(self) -> Tuple[Tuple[object, ...], Dict[str, object]]:
+        try:
+            next_sharding_choices = next(self.sharding_combs)
+            idx = 0
+
+            new_args: List[object] = []
+            for arg in self.flatten_args:
+                if isinstance(arg, torch.Tensor):
+                    new_args.append(
+                        self.to_dist_tensor(
+                            arg, self.mesh, [next_sharding_choices[idx]]
+                        )
+                    )
+                    idx += 1
+                else:
+                    new_args.append(arg)
+
+            new_kwargs: List[object] = []
+            for arg in self.flatten_kwargs:
+                if isinstance(arg, torch.Tensor):
+                    new_kwargs.append(
+                        self.to_dist_tensor(
+                            arg, self.mesh, [next_sharding_choices[idx]]
+                        )
+                    )
+                    idx += 1
+                else:
+                    new_kwargs.append(arg)
+
+            return tree_unflatten(
+                new_args, self.flatten_args_spec
+            ), tree_unflatten(new_kwargs, self.flatten_kwargs_spec)
+        except StopIteration:
+            raise StopIteration
+
+    def to_dist_tensor(
+        self,
+        t: torch.Tensor,
+        mesh: DeviceMesh,
+        placements: List[Placement],
+    ) -> torch.Tensor:
+        if type(t) is torch.Tensor or type(t) is torch.nn.Parameter:
+            if self.is_supported_tensor(t):
+                self.hit += 1
+                # We cannot use distribute_tensor for bool tensors as c10d
+                # collectives does not support the dtype, we assume op with
+                # bool tensor args the same tensor so we don't need to broadcast
+                # TODO: add bool tensor dtype support in c10d collective
+                if t.dtype == torch.bool:
+                    r = DTensor(
+                        t, mesh, placements, requires_grad=t.requires_grad
+                    )
+                else:
+                    r = distribute_tensor(t, mesh, placements)
+                if type(t) is torch.nn.Parameter:
+                    r = torch.nn.Parameter(r, requires_grad=r.requires_grad)  # type: ignore
+                return r
+            else:
+                self.miss += 1
+                return t
+        elif torch.overrides.is_tensor_like(t):
+            # Blindly converting tensor subclasses to dist tensor can cause
```

Comment:
Ideally we would just want to be able to have a DTensor of custom tensor, is this not allowed by design?